File: lib/iowa-mvp/guardrails.ts
/**
 * lib/iowa-mvp/guardrails.ts — Hard-coded safety guardrails for the Iowa MVP.
 *
 * Validates output text to ensure compliance with:
 * - No guarantee language
 * - No enrollment-advice language
 * - No discriminatory logic
 * - No unsupported claims
 */

// ---------------------------------------------------------------------------
// Forbidden phrase patterns
// ---------------------------------------------------------------------------

const GUARANTEE_PATTERNS = [
  /\bguaranteed?\b/i,
  /\bdefinitely covered\b/i,
  /\bdefinitely in[- ]network\b/i,
  /\byou will be approved\b/i,
  /\bexact cost will be\b/i,
  /\bwill definitely\b/i,
  /\b100% certain\b/i,
  /\bwe guarantee\b/i,
  /\bpromise you\b/i,
  /\bassured\b/i,
]

const ENROLLMENT_ADVICE_PATTERNS = [
  /\byou should enroll\b/i,
  /\bapply now\b/i,
  /\benroll now\b/i,
  /\bsign up (?:now|today|immediately)\b/i,
  /\bthis is the best plan for you\b/i,
  /\bthe best plan\b/i,
  /\bthe only plan you need\b/i,
  /\bwe recommend you choose\b/i,
  /\byou must select\b/i,
]

const UNSUPPORTED_CLAIM_PATTERNS = [
  /\byour doctor (?:is|will be) in[- ]network\b/i,
  /\byour pharmacy (?:is|will) cover\b/i,
  /\byour exact (?:premium|cost|price) (?:is|will be)\b/i,
  /\byou will (?:save|pay) exactly\b/i,
  /\byour subsidy (?:is|will be) exactly\b/i,
]

const DISCRIMINATORY_PATTERNS = [
  /\bpeople (?:like you|in your area|with your income) (?:should|deserve|only need)\b/i,
  /\bbased on your (?:neighborhood|zip code|race|gender|ethnicity)\b/i,
  /\byou (?:only )?deserve\b/i,
  /\blower[- ]income (?:people|individuals) should (?:accept|settle)\b/i,
]

// ---------------------------------------------------------------------------
// Validation functions
// ---------------------------------------------------------------------------

export interface GuardrailViolation {
  category: 'guarantee' | 'enrollment_advice' | 'unsupported_claim' | 'discriminatory'
  pattern: string
  matched_text: string
}

/**
 * Scan text for guardrail violations. Returns an array of violations found.
 * Empty array = text is safe.
 */
export function validateText(text: string): GuardrailViolation[] {
  const violations: GuardrailViolation[] = []

  const checkPatterns = (
    patterns: RegExp[],
    category: GuardrailViolation['category']
  ) => {
    for (const pattern of patterns) {
      const match = text.match(pattern)
      if (match) {
        violations.push({
          category,
          pattern: pattern.source,
          matched_text: match[0],
        })
      }
    }
  }

  checkPatterns(GUARANTEE_PATTERNS, 'guarantee')
  checkPatterns(ENROLLMENT_ADVICE_PATTERNS, 'enrollment_advice')
  checkPatterns(UNSUPPORTED_CLAIM_PATTERNS, 'unsupported_claim')
  checkPatterns(DISCRIMINATORY_PATTERNS, 'discriminatory')

  return violations
}

/**
 * Check if text is safe (no violations).
 */
export function isTextSafe(text: string): boolean {
  return validateText(text).length === 0
}

/**
 * Sanitize an array of explanation strings, removing any that contain violations.
 * Returns only safe strings.
 */
export function sanitizeExplanations(explanations: string[]): string[] {
  return explanations.filter((text) => isTextSafe(text))
}

// ---------------------------------------------------------------------------
// Required disclaimers
// ---------------------------------------------------------------------------

export const DISCLAIMERS = {
  general:
    'This tool provides estimates for educational purposes only. It is not insurance advice, and results are not a guarantee of coverage, costs, or eligibility. Plans, premiums, and benefits may differ from what is shown here.',
  non_enrollment:
    'This tool does not sell, solicit, or negotiate insurance. It is not affiliated with any government agency. Talk to a licensed agent or visit HealthCare.gov before making enrollment decisions.',
  data_snapshot:
    'Results are based on 2026 federal marketplace plan data loaded into this system. Carrier rules, plan availability, formulary coverage, and provider networks can change. Always verify details directly with the carrier or HealthCare.gov before enrolling.',
  subsidy_estimate:
    'Subsidy estimates shown here are approximate. Your actual premium tax credit depends on your final tax return, household composition, and the benchmark Silver plan in your area. Use the official HealthCare.gov calculator for binding estimates.',
  drug_coverage:
    'Drug coverage information is based on available formulary data. Not all carriers have verified formulary data in this system. Prior authorization, step therapy, and quantity limit requirements may apply. Confirm coverage with the carrier before enrolling.',
  no_provider_data:
    'This tool does not include provider directory or network data. Verify that your doctors, hospitals, and preferred pharmacies participate in any plan before enrolling.',
  non_discrimination:
    'This tool uses only lawful rating factors (age, location, tobacco use) and user-stated preferences. It does not discriminate based on gender, race, ethnicity, religion, disability, or any protected characteristic.',
} as const

export type DisclaimerKey = keyof typeof DISCLAIMERS

/**
 * Returns the standard set of disclaimers for display on the results page.
 */
export function getRequiredDisclaimers(): string[] {
  return [
    DISCLAIMERS.general,
    DISCLAIMERS.non_enrollment,
    DISCLAIMERS.data_snapshot,
    DISCLAIMERS.subsidy_estimate,
    DISCLAIMERS.no_provider_data,
  ]
}

File: scripts/generate/iowa_mvp/guardrails.py
"""
iowa_mvp/guardrails.py — Python reference guardrails for the Iowa MVP.

Mirrors the TypeScript guardrails in lib/iowa-mvp/guardrails.ts.
"""

import re
from dataclasses import dataclass

GUARANTEE_PATTERNS = [
    re.compile(r"\bguaranteed?\b", re.I),
    re.compile(r"\bdefinitely covered\b", re.I),
    re.compile(r"\bdefinitely in[- ]network\b", re.I),
    re.compile(r"\byou will be approved\b", re.I),
    re.compile(r"\bexact cost will be\b", re.I),
    re.compile(r"\bwill definitely\b", re.I),
    re.compile(r"\b100% certain\b", re.I),
    re.compile(r"\bwe guarantee\b", re.I),
    re.compile(r"\bpromise you\b", re.I),
    re.compile(r"\bassured\b", re.I),
]

ENROLLMENT_ADVICE_PATTERNS = [
    re.compile(r"\byou should enroll\b", re.I),
    re.compile(r"\bapply now\b", re.I),
    re.compile(r"\benroll now\b", re.I),
    re.compile(r"\bsign up (?:now|today|immediately)\b", re.I),
    re.compile(r"\bthis is the best plan for you\b", re.I),
    re.compile(r"\bthe best plan\b", re.I),
    re.compile(r"\bthe only plan you need\b", re.I),
    re.compile(r"\bwe recommend you choose\b", re.I),
    re.compile(r"\byou must select\b", re.I),
]

UNSUPPORTED_CLAIM_PATTERNS = [
    re.compile(r"\byour doctor (?:is|will be) in[- ]network\b", re.I),
    re.compile(r"\byour pharmacy (?:is|will) cover\b", re.I),
    re.compile(r"\byour exact (?:premium|cost|price) (?:is|will be)\b", re.I),
    re.compile(r"\byou will (?:save|pay) exactly\b", re.I),
    re.compile(r"\byour subsidy (?:is|will be) exactly\b", re.I),
]

DISCRIMINATORY_PATTERNS = [
    re.compile(r"\bpeople (?:like you|in your area|with your income) (?:should|deserve|only need)\b", re.I),
    re.compile(r"\bbased on your (?:neighborhood|zip code|race|gender|ethnicity)\b", re.I),
    re.compile(r"\byou (?:only )?deserve\b", re.I),
    re.compile(r"\blower[- ]income (?:people|individuals) should (?:accept|settle)\b", re.I),
]


@dataclass
class Violation:
    category: str
    pattern: str
    matched_text: str


def validate_text(text: str) -> list[Violation]:
    """Check text for guardrail violations. Returns list of violations."""
    violations: list[Violation] = []

    def check(patterns: list[re.Pattern[str]], category: str) -> None:
        for p in patterns:
            m = p.search(text)
            if m:
                violations.append(Violation(category, p.pattern, m.group()))

    check(GUARANTEE_PATTERNS, "guarantee")
    check(ENROLLMENT_ADVICE_PATTERNS, "enrollment_advice")
    check(UNSUPPORTED_CLAIM_PATTERNS, "unsupported_claim")
    check(DISCRIMINATORY_PATTERNS, "discriminatory")

    return violations


def is_text_safe(text: str) -> bool:
    """Return True if text passes all guardrails."""
    return len(validate_text(text)) == 0
